crates/bevy_ui/src/layout/ui_surface.rs

```diff
@@ -143,76 +181,150 @@ impl UiSurface {
-                        .entry(entity)
-                        .or_insert_with(|| self.taffy.new_leaf(viewport_style.clone()).unwrap());
-                    self.taffy.add_child(viewport_node, node).unwrap();
-
-                    RootNodePair {
-                        implicit_viewport_node: viewport_node,
-                        user_root_node: node,
-                    }
-                });
-            new_roots.push(root_node);
+        let removed_children = self.camera_root_nodes.entry(camera_entity).or_default();
+        let mut removed_children = removed_children.clone();
+
+        for ui_entity in children {
+            // creates mutable borrow on self that lives as long as the result
+            let _ = self.create_or_update_root_node_data(ui_entity, camera_entity);
+
+            // drop the mutable borrow on self by re-fetching
+            let root_node_data = self
+                .root_node_data
+                .get(&ui_entity)
+                .unwrap_or_else(|| unreachable!());
+
+            // fix taffy relationships
+            {
+                let taffy_node = *self.entity_to_taffy.get(&ui_entity).unwrap();
+                if let Some(parent) = self.taffy.parent(taffy_node) {
+                    self.taffy.remove_child(parent, taffy_node).unwrap();
+                }
+                self.taffy
+                    .add_child(root_node_data.implicit_viewport_node, taffy_node)
+                    .unwrap();
+            }
+            removed_children.remove(&ui_entity);
         }
 
-        self.camera_roots.insert(camera_id, new_roots);
+        for &orphan in removed_children.iter() {
+            self.mark_root_node_as_orphaned(orphan);
+        }
     }
 
     /// Compute the layout for each window entity's corresponding root node in the layout.
     pub fn compute_camera_layout<'a>(
         &mut self,
-        camera: Entity,
+        camera_entity: Entity,
         render_target_resolution: UVec2,
         buffer_query: &'a mut bevy_ecs::prelude::Query<&mut bevy_text::ComputedTextBlock>,
         font_system: &'a mut CosmicFontSystem,
     ) {
-        let Some(camera_root_nodes) = self.camera_roots.get(&camera) else {
+        let Some(camera_root_nodes) = self.camera_root_nodes.get(&camera_entity) else {
             return;
         };
 
         let available_space = taffy::geometry::Size {
             width: taffy::style::AvailableSpace::Definite(render_target_resolution.x as f32),
             height: taffy::style::AvailableSpace::Definite(render_target_resolution.y as f32),
         };
-        for root_nodes in camera_root_nodes {
+
+        for root_node_entity in camera_root_nodes {
+            let root_node_data = self
+                .root_node_data
+                .get(root_node_entity)
+                .expect("root_node_data missing");
+
+            if root_node_data.camera_entity.is_none() {
+                panic!("internal map out of sync");
```

Comment:
If I recall correctly, this is only possible because we expose several methods to `pub(super)` that could be used arbitrarily in the future and break the state. [I'll need to update this one, too, but the idea with this effort is an attempt to make the state more atomic](https://github.com/bevyengine/bevy/pull/13359)